File: Plugin.TokenCreator/TokenCreatorPlugin.cs
#region

using Eldora.PluginAPI;
using Newtonsoft.Json;
using Serilog;

#endregion

namespace Plugin.TokenCreator;

// TODO: Fix performance

public sealed class TokenCreatorPlugin : AbstractPlugin
{
	public static TokenCreatorPlugin Instance { get; private set; } = null!;

	public override List<IComponent> GetComponents()
	{
		return new List<IComponent>
		{
				new TokenCreatorPage()
		};
	}

	public string BordersImagesPath { get; private set; } = null!;
	private const string TokenBordersPrefixKey = "DefaultBorders";

	private static readonly List<string> DefaultTokens = new()
	{
			"10_sided_ring_1",
			"12_sided_ring_1",
			"abstract_double_ring_1",
			"abstract_double_ring_1_flat",
			"abstract_octagon_1",
			"abstract_octagon_1_flat",
			"abstract_ring_1",
			"abstract_ring_1_flat",
			"abstract_ring_2",
			"abstract_ring_2_flat",
			"abstract_ring_3",
			"abstract_ring_3_flat",
			"abstract_triangle_1",
			"cloud_ring_1",
			"double_ring_1",
			"fifthed_border_medium",
			"fifthed_border_medium_greyscale",
			"plain_hexagon_1",
			"plain_octagon_1",
			"plain_ring_1",
			"plain_ring_1_with_arrow_1",
			"plain_ring_2",
			"plain_ring_3",
			"plain_square_1",
			"plain_square_2",
			"spike_ring_1",
			"spike_ring_2",
			"wavey_ring_1"
	};

	private static readonly List<string> DefaultMasks = new()
	{
			"10_sided_ring_mask",
			"12_sided_ring_mask",
			"abstract_triangle_mask",
			"cloud_ring_1_mask",
			"plain_hexagon_mask",
			"plain_octagon_mask",
			"plain_ring_mask",
			"plain_square_mask",
			"wavey_ring_1_mask"
	};

	public readonly List<BorderObject> Borders = new();

	public override void OnLoad()
	{
		Instance = this;

		BordersImagesPath = Path.Join(PluginFolder, "borders");

		if (!Directory.Exists(BordersImagesPath))
		{
			Directory.CreateDirectory(BordersImagesPath);
		}

		SaveDefaultBorders();
		LoadBorders();
	}

	/// <summary>
	///     Loads all borders from the folder.
	/// </summary>
	private void LoadBorders()
	{
		foreach (var file in Directory.GetFiles(BordersImagesPath, "*.json"))
			try
			{
				var border = JsonConvert.DeserializeObject<BorderObject>(File.ReadAllText(file));

				if (border == null)
				{
					Log.Error("Failed to load border {File}", file);
					continue;
				}

				if (!File.Exists(Path.Combine(BordersImagesPath, border.ImageName)))
				{
					Log.Error("Image {ImageName} not found ({Path})",
							border.ImageName,
							Path.Combine(BordersImagesPath, border.ImageName));
					continue;
				}

				if (!File.Exists(Path.Combine(BordersImagesPath, border.MaskName)))
				{
					Log.Error("Mask {MaskName} not found ({Path})",
							border.MaskName,
							Path.Combine(BordersImagesPath, border.MaskName));
					continue;
				}

				Borders.Add(border);

				Log.Information("Loaded border {BorderName} with (ImageName: {ImageName} - MaskName: {MaskName})",
						border.BorderName,
						border.ImageName,
						border.MaskName);
			}
			catch (JsonSerializationException e)
			{
				Log.Warning(e, "Failed to load border {File}", file);
			}
	}

	/// <summary>
	///     Saves the default borders to plugin path
	/// </summary>
	private void SaveDefaultBorders()
	{
		foreach (var defaultToken in DefaultTokens)
		{
			var jsonName = $"{defaultToken}.json";
			var imageName = $"{defaultToken}.png";

			var externalJsonPath = Path.Combine(BordersImagesPath, jsonName);
			var externalImagePath = Path.Combine(BordersImagesPath, imageName);

			Utils.WriteResourceToFileIfNotExists($"{TokenBordersPrefixKey}.{jsonName}",
					externalJsonPath);

			Utils.WriteResourceToFileIfNotExists($"{TokenBordersPrefixKey}.{imageName}",
					externalImagePath);
		}

		foreach (var imageName in DefaultMasks.Select(mask => $"{mask}.png"))
		{
			var externalPath = Path.Combine(BordersImagesPath, $"{imageName}");

			Utils.WriteResourceToFileIfNotExists($"{TokenBordersPrefixKey}.{imageName}",
					externalPath);
		}

	}

}

File: TokenCreatorPlugin/src/main/java/io/ra6/eldora/tokencreator/TokenCreatorPlugin.java
package io.ra6.eldora.tokencreator;

import io.ra6.eldora.AbstractEldoraPlugin;
import org.apache.logging.log4j.core.util.FileUtils;
import org.apache.logging.log4j.core.util.IOUtils;
import org.json.JSONObject;

import java.io.*;
import java.nio.file.CopyOption;
import java.nio.file.Files;
import java.nio.file.StandardCopyOption;
import java.util.ArrayList;

public final class TokenCreatorPlugin extends AbstractEldoraPlugin {
	private static TokenCreatorPlugin INSTANCE;

	public static TokenCreatorPlugin getInstance() {
		return INSTANCE;
	}


	private final ArrayList<String> _defaultTokens = new ArrayList<>() {
		{
			add("10_sided_ring_1");
			add("12_sided_ring_1");
			add("abstract_double_ring_1");
			add("abstract_double_ring_1_flat");
			add("abstract_octagon_1");
			add("abstract_octagon_1_flat");
			add("abstract_ring_1");
			add("abstract_ring_1_flat");
			add("abstract_ring_2");
			add("abstract_ring_2_flat");
			add("abstract_ring_3");
			add("abstract_ring_3_flat");
			add("abstract_triangle_1");
			add("cloud_ring_1");
			add("double_ring_1");
			add("fifthed_border_medium");
			add("fifthed_border_medium_greyscale");
			add("plain_hexagon_1");
			add("plain_octagon_1");
			add("plain_ring_1");
			add("plain_ring_1_with_arrow_1");
			add("plain_ring_2");
			add("plain_ring_3");
			add("plain_square_1");
			add("plain_square_2");
			add("spike_ring_1");
			add("spike_ring_2");
			add("wavey_ring_1");
		}
	};
	private final ArrayList<String> _defaultMasks = new ArrayList<>() {
		{
			add("10_sided_ring_mask");
			add("12_sided_ring_mask");
			add("abstract_triangle_mask");
			add("cloud_ring_1_mask");
			add("plain_hexagon_mask");
			add("plain_octagon_mask");
			add("plain_ring_mask");
			add("plain_square_mask");
			add("wavey_ring_1_mask");
		}
	};

	private final ArrayList<BorderObject> _loadedBorders = new ArrayList<>();

	private String _pluginPath = "";
	private String _borderPath = "";

	@Override
	public void onLoad(String pluginPath) {
		INSTANCE = this;
		_pluginPath = pluginPath;
		_borderPath = _pluginPath + "/borders";

		addTabComponent(new TokenCreatorTab());

		exportDefaultBorders();
	}

	public String getBorderPath() {
		return _borderPath;
	}

	@Override
	public void onEnable() {
		loadBorders();
	}

	public BorderObject getBorderFromName(String name) {
		for (BorderObject borderObject : _loadedBorders) {
			if (borderObject.getName().equals(name)) {
				return borderObject;
			}
		}
		return null;
	}

	public ArrayList<BorderObject> getLoadedBorders() {
		return _loadedBorders;
	}

	private void loadBorders() {
		var files = new File(_borderPath).listFiles((dir, name) -> name.endsWith(".json"));

		if (files == null) return;

		for (var file : files) {
			try {
				loadBorder(file);
			} catch (IOException e) {
				getLogger().info(e);
			}
		}
	}

	private void loadBorder(File file) throws IOException {
		getLogger().info("Loading border {}", file.getName());

		var reader = new BufferedReader(new FileReader(file));
		var result = new StringBuilder();

		String line;
		while ((line = reader.readLine()) != null) {
			result.append(line);
		}
		reader.close();

		var plugin = new JSONObject(result.substring(result.indexOf("{")));
		if (!plugin.has("name")) {
			getLogger().error("Missing key {} on {}", "name", file.getName());
			return;
		}

		if (!plugin.has("image")) {
			getLogger().error("Missing key {} on {}", "image", file.getName());
			return;
		}

		if (!plugin.has("mask")) {
			getLogger().error("Missing key {} on {}", "mask", file.getName());
			return;
		}

		var border = new BorderObject(plugin.getString("name"), plugin.getString("image"), plugin.getString("mask"));
		_loadedBorders.add(border);
	}

	private void exportDefaultBorders() {
		_defaultTokens.forEach(tokenName -> {
			var pngPath = "borders/" + tokenName + ".png";
			var jsonPath = "borders/" + tokenName + ".json";
			export(pngPath, _pluginPath + "/" + pngPath);
			export(jsonPath, _pluginPath + "/" + jsonPath);

		});

		_defaultMasks.forEach(maskName -> {
			var path = "borders/" + maskName + ".png";
			export(path, _pluginPath + "/" + path);
		});
	}

	private void export(String internalPath, String filePath) {
		try (var stream = Thread.currentThread().getContextClassLoader().getResourceAsStream(internalPath)) {
			if (stream == null) return;

			var file = new File(filePath);
			if (!file.exists()) {
				//noinspection ResultOfMethodCallIgnored
				file.getParentFile().mkdir();
				getLogger().info("Copying {} to {}", internalPath, filePath);
				if (!file.createNewFile()) {
					getLogger().error("Could not create {}", filePath);
				}
			} else {
				getLogger().info("{} already exists. SKIPPING", filePath);
				return;
			}

			Files.copy(stream, file.toPath(), StandardCopyOption.REPLACE_EXISTING);
		} catch (IOException e) {
			throw new RuntimeException(e);
		}
	}
}
